File: app/src/main/java/me/gabeg/sicksends/SsProblemDao.java
package me.gabeg.sicksends;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.Query;
import androidx.room.Update;
import java.util.List;

/**
 * Data access object for climbing problems.
 */
@Dao
public abstract interface SsProblemDao<T>
{

	/**
	 * Delete a problem.
	 *
	 * @param  problem  Climbing problem.
	 */
	@Delete
	void delete(T problem);

	/**
	 * Delete all problems.
	 */
	//abstract void deleteAll();

	/**
	 * Insert a new problem.
	 *
	 * @param  problem  Climbing problem.
	 */
	@Insert
	void insert(T problem);

	/**
	 * Update an existing problem.
	 *
	 * @param  problem  Climbing problem.
	 */
	@Update
	void update(T problem);

	/**
	 * @return All climbing problems.
	 */
	abstract LiveData<List<T>> getAll();

}

File: app/src/main/java/me/gabeg/sicksends/problem/SsProblemDao.kt
package me.gabeg.sicksends.problem

import androidx.room.Dao
import androidx.room.Delete
import androidx.lifecycle.LiveData
import androidx.room.Insert
import androidx.room.Update
import kotlinx.coroutines.flow.Flow

/**
 * Data access object for climbing problems.
 */
@Dao
interface SsProblemDao<T : SsProblem>
{

	/**
	 * Delete a problem.
	 *
	 * @param  problem  Climbing problem.
	 */
	@Delete
	fun delete(problem: T)

	/**
	 * Delete all problems.
	 */
	abstract fun deleteAll()

	/**
	 * Insert a new problem.
	 *
	 * @param  problem  Climbing problem.
	 */
	@Insert
	suspend fun insert(problem: T)

	/**
	 * Update an existing problem.
	 *
	 * @param  problem  Climbing problem.
	 */
	@Update
	fun update(problem: T)

	/**
	 * Get all climbing problems.
	 *
	 * @return All climbing problems.
	 */
	val all: Flow<List<T>>
	//val all: Flow<List<T>?>?

}
